File: 2001-2500/2115_findAllRecipes.py
class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        graph = {supply: [] for supply in supplies}
        in_degree = {supply: 0 for supply in supplies}

        for i, ingredientList in enumerate(ingredients):
            for ingredient in ingredientList:
                if ingredient not in graph:
                    graph[ingredient] = []
                graph[ingredient].append(recipes[i])
                in_degree[recipes[i]] = in_degree.get(recipes[i], 0) + 1

        queue = deque(supplies)
        output = []

        while queue:
            supply = queue.popleft()
            if supply in graph:
                for ingredient in graph[supply]:
                    in_degree[ingredient] -= 1
                    if in_degree[ingredient] == 0:
                        output.append(ingredient)
                        queue.append(ingredient)
        
        return output

# Alternative solution

class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        # construct directed graph, direction from ingredients to recipes
        # construct indegree to determine whether recipe is reached
        graph = defaultdict(list)
        indegree = defaultdict(int)

        for i, recipe in enumerate(recipes):
            for ingredient in ingredients[i]:
                graph[ingredient].append(recipe)
                indegree[recipe] += 1

        queue = deque(supplies)
        result_list = []

        while queue:
            supply = queue.popleft()

            if supply in graph:
                for recipe in graph[supply]:
                    indegree[recipe] -= 1

                    if indegree[recipe] == 0:
                        result_list.append(recipe)
                        queue.append(recipe)

        return result_list

File: 2001-2500/2115_findAllRecipes.java
class Solution {
    public List<String> findAllRecipes(String[] recipes, List<List<String>> ingredients, String[] supplies) {
        HashMap<String, List<String>> recipeGraph = new HashMap<>();
        HashMap<String, Integer> indegree = new HashMap<>();
        Queue<String> queue = new LinkedList<>();
        List<String> output = new ArrayList<>();

        for (int i = 0; i < recipes.length; i++) {
            if (!recipeGraph.containsKey(recipes[i])) {
                recipeGraph.put(recipes[i], new ArrayList<>());
            }
            
            for (String ingredient: ingredients.get(i)) {
                if (!recipeGraph.containsKey(ingredient)) {
                    recipeGraph.put(ingredient, new ArrayList<>());
                }
                recipeGraph.get(ingredient).add(recipes[i]);

                indegree.put(recipes[i], indegree.getOrDefault(recipes[i], 0) + 1);
            }
        }

        for (String supply: supplies) {
            if (recipeGraph.containsKey(supply) ) {
                queue.add(supply);
            }
        }

        while (!queue.isEmpty()) {
            String node = queue.remove();

            if (indegree.containsKey(node)) {
                output.add(node);
            }

            for (String child: recipeGraph.get(node)) {
                indegree.put(child, indegree.get(child) - 1);

                if (indegree.get(child) == 0) {
                    queue.add(child);
                }
            }
        }

        return output;
    }
}
